galsim/vonkarman.py

```diff
@@ -0,0 +1,183 @@
+#
+# GalSim is free software: redistribution and use in source and binary forms,
+# with or without modification, are permitted provided that the following
+# conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions, and the disclaimer given in the accompanying LICENSE
+#    file.
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions, and the disclaimer given in the documentation
+#    and/or other materials provided with the distribution.
+#
+"""@file base.py
+This file implements the von Karman atmospheric PSF profile.
+"""
+
+import numpy as np
+
+import galsim
+
+from . import _galsim
+from .gsobject import GSObject
+
+
+class VonKarman(GSObject):
+    """Class describing a von Karman surface brightness profile, which represents a long exposure
+    atmospheric PSF.  The difference between the von Karman profile and the related Kolmogorov
+    profile is that the von Karman profile includes a parameter for the outer scale of atmospheric
+    turbulence, which is a physical scale beyond which fluctuations in the refractive index stop
+    growing, typically between 10 and 100 meters.  Quantitatively, the von Karman phase fluctuation
+    power spectrum is proportional to
+
+        (f^2 + L0^-2)^(-11/6)
+
+    where f is a spatial frequency and L0 is the outer scale in meters.  The Kolmogorov power
+    spectrum proportional to f^(-11/3) is recovered as L0 -> infinity,
+
+    For more information, we recommend the following references:
+
+        Martinez et al.  2010  A&A  vol. 516
+        Conan  2008  JOSA  vol. 25
+
+    Notes
+    -----
+
+    If one blindly follows the math for converting the von Karman power spectrum into a PSF, one
+    finds that the PSF contains a delta-function at the origin with fractional flux of
+
+        exp(-0.5*0.172*(r0/L0)^(-5/3))
+
+    In almost all cases of interest this evaluates to something tiny, often on the order of 10^-100
+    or smaller.  By default, GalSim will ignore this delta function entirely.  If for some reason
+    you want to keep the delta function, though, then you can pass the do_delta=True argument to the
+    VonKarman initializer.
+
+    @param lam               Wavelength in nanometers
+    @param r0                Fried parameter in meters.
+    @param L0                Outer scale in meters.  [default: np.inf]
+    @param flux              The flux (in photons/cm^2/s) of the profile. [default: 1]
+    @param scale_unit        Units assumed when drawing this profile or evaluating xValue, kValue,
+                             etc.  Should be a galsim.AngleUnit or a string that can be used to
+                             construct one (e.g., 'arcsec', 'radians', etc.).
+                             [default: galsim.arcsec]
+    @param do_delta          Include delta-function at origin? (not recommended; see above).
```

Comment:
Not recommended just because it's usually so small?  Or does it cause some numerical problem?  If the latter, should say this (because so far you've only mentioned the smallness issue).

Reply:
There are some numerical problems which led, for example, to restricting some of the unit tests to only operate on profiles without the delta function.  I modified some of the preceding text to say: 

```
    By default, GalSim will ignore this delta function entirely since it usually
    doesn't make any difference, but can complicate some calculations like drawing using
    method='real_space' or by formally requiring huge Fourier transforms for drawing using 
    method='fft'.
```